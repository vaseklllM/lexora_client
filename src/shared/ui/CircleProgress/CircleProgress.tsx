import { CSSProperties, ReactElement } from "react";
import { tv } from "tailwind-variants";

const classesSlots = tv({
  slots: {
    base: "radial-progress bg-base-200 text-base-content border-base-300/0 border-2 text-sm",
  },
});

interface Props {
  className?: string;
  value?: number;
}

export const CircleProgress = (props: Props): ReactElement | null => {
  const classes = classesSlots();

  if (typeof props.value !== "number") {
    return null;
  }

  return (
    <div
      style={
        {
          "--value": props.value,
          "--size": "30px",
          "--thickness": "2px",
        } as CSSProperties
      }
      aria-valuenow={props.value}
      role="progressbar"
      className={classes.base({ className: props.className })}
    >
      {Math.round(props.value / 10)}
    </div>
  );
};
