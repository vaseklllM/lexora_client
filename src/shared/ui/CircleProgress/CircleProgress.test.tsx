import { render, screen } from "@testing-library/react";
import { CircleProgress } from "./CircleProgress";

describe("CircleProgress", () => {
  it("should return null when value is not a number", () => {
    const { container } = render(<CircleProgress value={undefined} />);
    expect(container.firstChild).toBeNull();
  });

  it("should render with given value", () => {
    render(<CircleProgress value={50} />);

    const progressbar = screen.getByRole("progressbar");
    expect(progressbar).toBeInTheDocument();
  });

  it("should display correct rounded value", () => {
    render(<CircleProgress value={75} />);

    const progressbar = screen.getByRole("progressbar");

    expect(progressbar).toHaveTextContent("8");
  });

  it("should have correct aria-valuenow attribute", () => {
    render(<CircleProgress value={30} />);

    const progressbar = screen.getByRole("progressbar");
    expect(progressbar).toHaveAttribute("aria-valuenow", "30");
  });

  it("should render with zero value", () => {
    render(<CircleProgress value={0} />);

    const progressbar = screen.getByRole("progressbar");
    expect(progressbar).toHaveTextContent("0");
    expect(progressbar).toHaveAttribute("aria-valuenow", "0");
  });

  it("should render with 100 value", () => {
    render(<CircleProgress value={100} />);

    const progressbar = screen.getByRole("progressbar");
    expect(progressbar).toHaveTextContent("10");
    expect(progressbar).toHaveAttribute("aria-valuenow", "100");
  });

  it("should apply custom className", () => {
    render(<CircleProgress value={50} className="custom-class" />);

    const progressbar = screen.getByRole("progressbar");
    expect(progressbar).toHaveClass("custom-class");
  });

  it("should have correct CSS custom properties", () => {
    render(<CircleProgress value={45} />);

    const progressbar = screen.getByRole("progressbar");
    const style = progressbar.style;

    expect(style.getPropertyValue("--value")).toBe("45");
    expect(style.getPropertyValue("--size")).toBe("30px");
    expect(style.getPropertyValue("--thickness")).toBe("2px");
  });

  it("should match snapshot with value 0", () => {
    const { container } = render(<CircleProgress value={0} />);
    expect(container).toMatchSnapshot();
  });

  it("should match snapshot with value 50", () => {
    const { container } = render(<CircleProgress value={50} />);
    expect(container).toMatchSnapshot();
  });

  it("should match snapshot with value 100", () => {
    const { container } = render(<CircleProgress value={100} />);
    expect(container).toMatchSnapshot();
  });
});
