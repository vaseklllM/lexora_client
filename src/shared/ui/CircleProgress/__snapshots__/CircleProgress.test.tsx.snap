// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircleProgress should match snapshot with value 0 1`] = `
<div>
  <div
    aria-valuenow="0"
    class="radial-progress bg-base-200 text-base-content border-base-300/0 border-2 text-sm"
    role="progressbar"
    style="--value: 0; --size: 30px; --thickness: 2px;"
  >
    0
  </div>
</div>
`;

exports[`CircleProgress should match snapshot with value 50 1`] = `
<div>
  <div
    aria-valuenow="50"
    class="radial-progress bg-base-200 text-base-content border-base-300/0 border-2 text-sm"
    role="progressbar"
    style="--value: 50; --size: 30px; --thickness: 2px;"
  >
    5
  </div>
</div>
`;

exports[`CircleProgress should match snapshot with value 100 1`] = `
<div>
  <div
    aria-valuenow="100"
    class="radial-progress bg-base-200 text-base-content border-base-300/0 border-2 text-sm"
    role="progressbar"
    style="--value: 100; --size: 30px; --thickness: 2px;"
  >
    10
  </div>
</div>
`;
